Src/Newtonsoft.Json/Utilities/MethodBinder.cs

```diff
@@ -0,0 +1,333 @@
+﻿namespace Newtonsoft.Json.Utilities
+{
+#if PORTABLE
+
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+
+    internal static class MethodBinder
+    {
+        
+        /// <summary>
+        /// List of primitive types which can be widened.
+        /// </summary>
+        private static readonly Type[] PrimitiveTypes = new Type[]
+        {
+            typeof(bool),  typeof(char),   typeof(sbyte), typeof(byte),
+            typeof(short), typeof(ushort), typeof(int),   typeof(uint),
+            typeof(long),  typeof(ulong),  typeof(float), typeof(double)
+        };
+
+        /// <summary>
+        /// Widening masks for primitive types above.
+        /// Index of the value in this array defines a type we're widening,
+        /// while the bits in mask define types it can be widened to (including itself).
+        /// 
+        /// For example, value at index 0 defines a bool type, and it only has bit 0 set, 
+        /// i.e. bool values can be assigned only to bool.
+        /// </summary>
+        private static readonly int[] WideningMasks = new int[]
+        {
+            0x0001,        0x0FE2,         0x0D54,        0x0FFA,
+            0x0D50,        0x0FE2,         0x0D40,        0x0F80,
+            0x0D00,        0x0E00,         0x0C00,        0x0800
+        };
+
+        /// <summary>
+        /// Checks if value of primitive type <paramref name="from"/> can be  
+        /// assigned to parameter of primitive type <paramref name="to"/>.
+        /// </summary>
+        /// <param name="from">Source primitive type</param>
+        /// <param name="to">Target primitive type</param>
+        /// <returns>True if source type can be widened to target type, false otherwise</returns>
+        internal static bool CanConvertPrimitive(Type from, Type to)
+        {
+            if (from == to)
+            {
+                // same type
+                return true;
+            }
+
+            int fromMask = 0, toMask = 0;
+
+            for (int i = 0; i < PrimitiveTypes.Length && (fromMask == 0 || toMask == 0); i++)
```

Comment:
Can you move the `(fromMask == 0 || toMask == 0)` to its own if statement inside the for loop?